File: Sources/PUPSISPortalApp/Core/ScheduleStore.swift
import Foundation

/// The schedule as it was last scraped, plus when that happened.
struct CachedSchedule: Codable, Equatable {
    let lastUpdated: Date
    let sessions: [ClassSession]
}

/// On-disk cache of the scraped schedule, so a launch can render before the
/// network answers — and keep rendering when it never does.
///
/// Application Support, not the Keychain (this isn't a secret) and not
/// `UserDefaults` (it's a document, not a preference).
enum ScheduleStore {
    static let fileURL: URL = {
        let support = (try? FileManager.default.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )) ?? URL(fileURLWithPath: NSTemporaryDirectory())
        return support
            .appendingPathComponent("PUPSISPortal", isDirectory: true)
            .appendingPathComponent("schedule.json")
    }()

    static func save(_ sessions: [ClassSession], lastUpdated: Date = Date(), to url: URL = fileURL) {
        let payload = CachedSchedule(lastUpdated: lastUpdated, sessions: sessions)
        guard let data = try? JSONEncoder().encode(payload) else { return }

        let directory = url.deletingLastPathComponent()
        try? FileManager.default.createDirectory(
            at: directory,
            withIntermediateDirectories: true,
            attributes: [.posixPermissions: 0o700]
        )
        guard (try? data.write(to: url, options: .atomic)) != nil else { return }
        // The user's own class schedule — readable by them, nobody else on
        // the machine. `.atomic` replaces the file, so re-apply every write.
        try? FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: url.path)
    }

    /// `nil` for "no usable cache". A missing, unreadable, or stale-format
    /// file is not an error worth surfacing — one refresh rebuilds it.
    static func load(from url: URL = fileURL) -> CachedSchedule? {
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONDecoder().decode(CachedSchedule.self, from: data)
    }

    static func delete(at url: URL = fileURL) {
        try? FileManager.default.removeItem(at: url)
    }
}

File: windows/PUPSISPortal.Core/ScheduleStore.cs
namespace PUPSISPortal.Core;

/// <summary>
/// On-disk cache of the scraped schedule, so a launch can render before the
/// network answers — and keep rendering when it never does.
///
/// The student's own class schedule: LocalAppData, directory with owner-only ACL,
/// cleared on sign-out.
/// </summary>
public static class ScheduleStore
{
    private const string FileName = "schedule.json";
    private static string? s_testDirectory = null;

    /// <summary>
    /// For testing: override the cache directory. Pass null to use the default.
    /// </summary>
    public static void SetTestDirectory(string? directory)
    {
        s_testDirectory = directory;
    }

    /// <summary>
    /// Save `sessions` and `lastUpdated` to the cache.
    /// Silently fails if serialization or I/O fails — no need to surface
    /// cache write failures to the UI.
    /// </summary>
    public static void Save(List<ClassSession> sessions, DateTime? lastUpdated = null)
    {
        var payload = new CachedSchedule
        {
            LastUpdated = lastUpdated ?? DateTime.UtcNow,
            Sessions = sessions
        };
        JsonStore.Save(payload, FileName, s_testDirectory);
    }

    /// <summary>
    /// Load the cached schedule from disk.
    /// Returns null if the file doesn't exist, can't be read, or fails to deserialize.
    /// A missing, unreadable, or stale-format file is not an error worth surfacing —
    /// one refresh rebuilds it.
    /// </summary>
    public static CachedSchedule? Load()
    {
        return JsonStore.Load<CachedSchedule>(FileName, s_testDirectory);
    }

    /// <summary>
    /// Delete the cache file from disk.
    /// Silently fails if the file doesn't exist or can't be deleted.
    /// </summary>
    public static void Delete()
    {
        JsonStore.Delete(FileName, s_testDirectory);
    }
}
